File: src/csharp/Intel/Iced.UnitTests/Intel/FormatterTests/TestSymbolResolver.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if GAS || INTEL || MASM || NASM || FAST_FMT
using System.Linq;
using Iced.Intel;

namespace Iced.UnitTests.Intel.FormatterTests {
	sealed class TestSymbolResolver : ISymbolResolver {
		readonly SymbolResolverTestCase tc;

		public TestSymbolResolver(in SymbolResolverTestCase tc) => this.tc = tc;

		public bool TryGetSymbol(in Instruction instruction, int operand, int instructionOperand, ulong address, int addressSize, out SymbolResult symbol) {
			foreach (var tc in tc.SymbolResults) {
				if (tc.Address != address || tc.AddressSize != addressSize)
					continue;
				var text = new TextInfo(tc.SymbolParts.Select(a => new TextPart(a, FormatterTextKind.Text)).ToArray());
				if (tc.MemorySize != null)
					symbol = new SymbolResult(tc.SymbolAddress, text, tc.Flags, tc.MemorySize.Value);
				else
					symbol = new SymbolResult(tc.SymbolAddress, text, tc.Flags);
				return true;
			}
			symbol = default;
			return false;
		}
	}
}
#endif

File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/fmt/TestSymbolResolver.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.fmt;

import com.github.icedland.iced.x86.Instruction;

final class TestSymbolResolver implements SymbolResolver {
	final SymbolResolverTestCase tc;

	public TestSymbolResolver(SymbolResolverTestCase tc) {
		this.tc = tc;
	}

	public SymbolResult getSymbol(Instruction instruction, int operand, int instructionOperand, long address, int addressSize) {
		for (SymbolResultTestCase tc : this.tc.symbolResults) {
			if (tc.address != address || tc.addressSize != addressSize)
				continue;
			TextPart[] parts = new TextPart[tc.symbolParts.length];
			for (int i = 0; i < parts.length; i++)
				parts[i] = new TextPart(tc.symbolParts[i], FormatterTextKind.TEXT);
			TextInfo text = new TextInfo(parts);
			if (tc.memorySize != null)
				return new SymbolResult(tc.symbolAddress, text, tc.flags, tc.memorySize);
			return new SymbolResult(tc.symbolAddress, text, tc.flags, true);
		}
		return null;
	}
}
